// evergreen-ui/commonjs/theme/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveThemeTokens = exports.mergeTheme = exports.getValue = exports.get = exports.useTheme = exports.withTheme = exports.ThemeConsumer = exports.ThemeProvider = exports.getThemeContext = exports.ThemeContext = void 0;
var ThemeContext_1 = require("./src/ThemeContext");
Object.defineProperty(exports, "ThemeContext", { enumerable: true, get: function () { return __importDefault(ThemeContext_1).default; } });
Object.defineProperty(exports, "getThemeContext", { enumerable: true, get: function () { return ThemeContext_1.getThemeContext; } });
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return ThemeContext_1.ThemeProvider; } });
Object.defineProperty(exports, "ThemeConsumer", { enumerable: true, get: function () { return ThemeContext_1.ThemeConsumer; } });
var withTheme_1 = require("./src/withTheme");
Object.defineProperty(exports, "withTheme", { enumerable: true, get: function () { return __importDefault(withTheme_1).default; } });
var useTheme_1 = require("./src/useTheme");
Object.defineProperty(exports, "useTheme", { enumerable: true, get: function () { return __importDefault(useTheme_1).default; } });
var theme_tools_1 = require("./src/theme-tools");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return theme_tools_1.get; } });
Object.defineProperty(exports, "getValue", { enumerable: true, get: function () { return theme_tools_1.getValue; } });
Object.defineProperty(exports, "mergeTheme", { enumerable: true, get: function () { return theme_tools_1.mergeTheme; } });
Object.defineProperty(exports, "resolveThemeTokens", { enumerable: true, get: function () { return theme_tools_1.resolveThemeTokens; } });
