// evergreen-ui/codemods/dist/utils/register-collection-extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCollectionExtensions = void 0;
const once_1 = require("./once");
const _registerCollectionExtensions = (jscodeshift) => {
    const j = jscodeshift;
    j.registerMethods({
        concat: function (rightCollection) {
            const thisCollection = this;
            return j([...thisCollection.paths(), ...rightCollection.paths()]);
        },
        difference: function (rightCollection) {
            const thisCollection = this;
            return thisCollection.filter(leftNode => rightCollection.every(rightNode => leftNode !== rightNode));
        },
        findImportDeclarationByModuleName: function (moduleName) {
            const thisCollection = this;
            return thisCollection.find(j.ImportDeclaration, {
                source: { value: moduleName }
            });
        },
        findEvergreenImportDeclaration: function () {
            const thisCollection = this;
            return thisCollection.findImportDeclarationByModuleName('evergreen-ui');
        },
        firstNode: function (predicate) {
            const thisCollection = this;
            const filteredCollection = predicate != null ? thisCollection.filter(predicate).at(0) : thisCollection.at(0);
            return filteredCollection.nodes()[0];
        },
        hasValues: function () {
            const thisCollection = this;
            return thisCollection.length > 0;
        },
        isEmpty: function () {
            const thisCollection = this;
            return thisCollection.length === 0;
        },
        toNodeArray: function () {
            const thisCollection = this;
            return thisCollection.nodes();
        }
    });
    return j;
};
const registerCollectionExtensions = (0, once_1.once)(_registerCollectionExtensions);
exports.registerCollectionExtensions = registerCollectionExtensions;
